### PR TITLE
[Bugfix] Fix ray instance detect issue

### PR DESCRIPTION
**Pull request by yma11**
Fix ray instance detect so that will first try connecting to latest launched instance and if not, create a new one with `num_gpus=parallel_config.world_size`.

### CODE REVIEW

File: vllm/executor/ray_utils.py [hunk 1]
```diff
@@ -229,12 +229,17 @@ def initialize_ray_cluster(
...
     if is_hip() or is_xpu():
-        ray.init(address=ray_address,
-                 ignore_reinit_error=True,
-                 num_gpus=parallel_config.world_size)
+        # Try to connect existing ray instance and create a new one if not found
+        try:
+            ray.init('auto')
+        except ray.exceptions.ConnectionError:
+            logger.warning("No existing RAY instance detected. " +
+                           "A new instance will be launched with current node resources.")
```
Comment (russellb) [inline vllm/executor/ray_utils.py:239]: @yma11 it looks like you need to run `format.sh` and you'll see that it's generating an error that this line (and one other) is too long. That's why CI is failing.

**Review by russellb — CHANGES_REQUESTED** [review]

File: vllm/executor/ray_utils.py (hunk 1 above)
Comment (yma11) [reply vllm/executor/ray_utils.py:239]: Thanks for comments. I will update this PR soon. Just working on fix some issues in my env to run `format.sh`.

**Comment by yma11** [comment]
@youkaichao can you help review this change? Thanks.

File: vllm/executor/ray_utils.py [hunk 2]
```diff
@@ -229,12 +229,18 @@ def initialize_ray_cluster(
...
     assert_ray_available()
-
     # Connect to a ray cluster.
     if is_hip() or is_xpu():
-        ray.init(address=ray_address,
-                 ignore_reinit_error=True,
-                 num_gpus=parallel_config.world_size)
+        # Try to connect existing ray instance and create a new one if not found
+        try:
+            ray.init('auto')
```
Comment (comaniac) [inline vllm/executor/ray_utils.py:236]: Use double quotes for consistency.

**Review by comaniac — COMMENTED** [review]
Overall LGTM. One question is should we also change the other branch?

**Comment by yma11** [comment]
> Overall LGTM. One question is should we also change the other branch?

Agree. So I unify the `init` logic which should make sense for all platforms, please help take a look again. Thanks.

**Comment by comaniac** [comment]
The update code seems not equivalent to the original one? Currently for non-hip and non-xpu cases we don't init Ray with all GPUs.

**Comment by yma11** [comment]
> The update code seems not equivalent to the original one? Currently for non-hip and non-xpu cases we don't init Ray with all GPUs.

For non-hip and non-xpu cases, it will create a local instance with detected gpus if fails to connect existing cluster based on [explanation](url).
Actually I intended to fix an error `When connecting to an existing cluster, num_cpus and num_gpus must not be provided.`  in xpu case. It happens when a valid `ray_address` and `num_gpus` are both given. I want to respect both of these values but seems the confliction can't be resolved. Maybe it's more reasonable to do ` ray.init(address=ray_address, ignore_reinit_error=True)` for all platforms. `num_gpus=parallel_config.world_size` is expected to take affect only when new local instance created but it's not so meaningful in that case. What do you think?

**Comment by comaniac** [comment]
Sounds reasonable to me, but cc @rkooo567 @richardliaw to double check.

**Comment by youkaichao** [comment]
@yma11 please resolve the conflict

**Review by rkooo567 — APPROVED** [review]

**Comment by yma11** [comment]
@youkaichao Thanks for reminder. @comaniac I switched the fix back to only change `hip` and `xpu` code path since there is an  possible issue on these platforms. When there is no ray cluster existing and trying to launch a new instance, Ray may can't detect correct GPU numbers thus will cause no GPU resources available for ray worker allocation. So we need give `num_gpus` as the argument in this case. That's why this specific code path exists here. FYI and thanks for your review.

**Comment by youkaichao** [comment]
@DarkLight1337 please help check is the error related or it occurs in the main branch previously?

**Comment by DarkLight1337** [comment]
It is a failure from main branch that has since been fixed. You can force merge this.

**Pull request merged**